File: Networking/Objects/NetworkObject.cs
using Common.Extensions;
using Common.IO.Collections;
using Common.Utilities;

using Networking.Features;

using System;
using System.Threading;

namespace Networking.Objects
{
    public class NetworkObject
    {
        private LockedDictionary<ushort, NetworkVariable> netFields;
        private Timer netTimer;
        private Type thisType;

        public bool isDestroyed;
        public bool isReady;

        public readonly NetworkManager manager;
        public readonly NetworkFunctions net;
        public readonly ushort typeHash;

        public NetworkObject(NetworkManager manager)
        {
            this.manager = manager;
            this.net = manager.net;
            this.thisType = GetType();
            this.typeHash = this.thisType.GetTypeHash();
            this.netFields = new LockedDictionary<ushort, NetworkVariable>();
        }

        public virtual void OnStart() { }
        public virtual void OnStop() { }

        public void SendRpc(ushort functionHash, params object[] args)
            => net.Send(new NetworkRpcMessage(typeHash, functionHash, args));

        public void SendCmd(ushort functionHash, params object[] args)
            => net.Send(new NetworkCmdMessage(typeHash, functionHash, args));

        public void SendEvent(ushort eventHash, bool shouldExecute, params object[] args)
        {
            if (shouldExecute && manager.netEvents.TryGetValue(eventHash, out var ev))
                ev.Raise(this, args);

            net.Send(new NetworkRaiseEventMessage(typeHash, eventHash, args));
        }

        internal void StartInternal()
        {
            foreach (var fieldPair in manager.netFields)
            {
                if (fieldPair.Value.DeclaringType != thisType)
                    continue;

                var fieldNetVar = fieldPair.Value.FieldType.Construct() as NetworkVariable;

                fieldNetVar.parent = this;
                fieldPair.Value.SetValueFast(fieldNetVar, this);

                netFields[fieldPair.Key] = fieldNetVar;
            }

            CodeUtils.WhileTrue(() => !isDestroyed && isReady, UpdateNetFields, 10);
        }

        internal void StopInternal()
        {
            netTimer?.Dispose();
            netTimer = null;

            netFields?.Clear();
            netFields = null;
        }

        internal void ProcessVarSync(NetworkVariableSyncMessage syncMsg)
        {
            if (!netFields.TryGetValue(syncMsg.hash, out var netVar))
            {
                manager.log.Warn($"Received a sync message for an unknown network variable: {syncMsg.hash}");
                return;
            }

            netVar.Process(syncMsg.msg);
        }

        private void UpdateNetFields()
        {
            foreach (var netField in netFields)
            {
                while (netField.Value.pending.TryDequeue(out var syncMsg))
                    net.Send(new NetworkVariableSyncMessage(typeHash, netField.Key, syncMsg));
            }
        }
    }
}

File: Networking/Objects/NetworkObject.vb
Imports Common.Extensions
Imports Common.IO.Collections

Imports Networking.Features

Imports System
Imports System.Threading

Namespace Networking.Objects
    Public Class NetworkObject
        Private netFields As LockedDictionary(Of UShort, NetworkVariable)
        Private netTimer As Timer
        Private thisType As Type

        Public isDestroyed As Boolean
        Public isReady As Boolean

        Public ReadOnly id As Integer

        Public ReadOnly manager As NetworkManager
        Public ReadOnly net As NetworkFunctions

        Public Sub New(id As Integer, manager As NetworkManager)
            Me.id = id
            Me.manager = manager
            net = manager.net
            thisType = [GetType]()
        End Sub

        Public Overridable Sub OnStart()
        End Sub
        Public Overridable Sub OnStop()
        End Sub

        Public Sub Destroy()
            manager.Destroy(Me)
        End Sub

        Public Sub SendRpc(functionHash As UShort, ParamArray args As Object())
            net.Send(New NetworkRpcMessage(id, functionHash, args))
        End Sub

        Public Sub SendCmd(functionHash As UShort, ParamArray args As Object())
            net.Send(New NetworkCmdMessage(id, functionHash, args))
        End Sub

        Friend Sub StartInternal()
            For Each fieldPair In manager.netFields
                If fieldPair.Value.DeclaringType IsNot thisType Then Continue For

                Dim fieldNetVar = TryCast(fieldPair.Value.FieldType.Construct(), NetworkVariable)

                fieldNetVar.parent = Me

                fieldPair.Value.SetValueFast(Me, fieldNetVar)

                netFields(fieldPair.Key) = fieldNetVar
            Next

            netTimer = New Timer(Sub(__) UpdateNetFields(), Nothing, 100, 100)
        End Sub

        Friend Sub StopInternal()
            netTimer?.Dispose()
            netTimer = Nothing

            netFields?.Clear()
            netFields = Nothing
        End Sub

        Friend Sub ProcessVarSync(syncMsg As NetworkVariableSyncMessage)
            Dim field As FieldInfo = Nothing, netVar As NetworkVariable = Nothing
            If Not manager.netFields.TryGetValue(syncMsg.hash, field) OrElse field.DeclaringType IsNot thisType OrElse Not netFields.TryGetValue(syncMsg.hash, netVar) Then Return

            netVar.Process(syncMsg.msg)
        End Sub

        Private Sub UpdateNetFields()
            For Each netField In netFields
                While netField.Value.pending.Count > 0
                    Dim msg = netField.Value.pending(0)

                    netField.Value.pending.RemoveAt(0)

                    net.Send(New NetworkVariableSyncMessage(id, netField.Key, msg))
                End While
            Next
        End Sub
    End Class
End Namespace
